---
title: "Create phyloseq object of zOTU dataset"
author: "Madeleine Gundersen"
date: "21 06 2020"
---

Script for making a phyloseq object from an zOTU table. The phyloseq class merges information about your dataset, and incorporates OTU tables, taxonomy tables, phylogenecy and sample classes. The phyloseq-class is intended to allow the representation of all the useful features of a phylogenetic sequencing project as one coherent object (see https://github.com/joey711/phyloseq/wiki/phyloseq-class-structure-(developer) ) 


```{r}
library("phyloseq") 
library("readxl")
library("magrittr")
library("ape")
library(genefilter) #BiocManager::install("genefilter")
```

```{r}
set.seed(0201)
sessionInfo()
filepath = "~/OneDrive - NTNU/Doktorgrad/Project 3 MMS_FTS_r_K_experiment/MMS-FTS Overview results and manuscript plan/R_Analysis_MMS-FTS/"
setwd(filepath)
```

2. Import your OTU table
```{r}
##### How to load from excel file
otumatrix = readxl::read_xlsx(paste0(filepath, "/data/zOTUs/zOTUs_withnames.xlsx"), sheet = 1) %>% as.data.frame
# changes the rownames to OTU ids
row.names(otumatrix) = otumatrix[, 1]
# deletes the first column
otumatrix = subset(otumatrix, select = c(colnames(otumatrix)[2:length(otumatrix)]))
# makes a phyloseq object with an OTU table
OTU = phyloseq::otu_table(otumatrix, taxa_are_rows = TRUE)
```

2. Import taxonomy information
import_sintax_file.R can be found and downloaded here: https://rdrr.io/github/jfq3/RDPutils/src/R/import_sintax_file.R

```{r}
source(paste0(filepath, "/functions/import_sintax-file.R") ) 
taxmat = import_sintax_file(paste0(filepath, "/data/zotus/zotus.sintax"), confidence = 0.8)
TAX = phyloseq::tax_table(taxmat)
experiment = phyloseq::phyloseq(OTU, TAX)
```

3. Import variables 
```{r}
variables = readxl::read_xlsx(paste0(filepath, "data/zotus/2021-06-21_zOTU_metadata.xlsx" ) , 
                              sheet = 1) %>% as.data.frame
#changes the row names to the sample names. These must be the same as the sample names in the OTU file
row.names(variables)=variables$`Sample`  #Sample names is the column name of the samples 
sam_variables = phyloseq::sample_data(variables)
experiment = phyloseq::merge_phyloseq(experiment, sam_variables)
```

4. Import phylogenetic tree 
```{r}
#load phylogenetic tree created previously 
tree.unroot_whole = readRDS(file = paste0(filepath, "data/zotus/zOTU_phytree.RDS")  ) 
tree.unrooted = tree.unroot_whole$tree

#remove ASVs that are biased (chloroplasts are alginates, Zotu3717 human and Zotu2353 most likely algea)
chloroplasts = subset_taxa(physeq =experiment, Class == "c_Chloroplast")
badTaxa = c("Zotu3717", "Zotu2353" , taxa_names(chloroplasts)) 
goodTaxa <- setdiff(taxa_names(experiment), badTaxa)
experiment <- prune_taxa(goodTaxa, experiment)
experiment = merge_phyloseq(experiment, tree.unrooted) #dataset contains 2415 taxa after removing biased sequences
#remove all ASVs with less than 8 reads total
#
# This is not pokay :) already removed
# experiment_8 = filter_taxa(physeq = experiment,flist =  function(x) sum(x) > 8,prune =  TRUE) #dataset contains 1316 taxa

#root phylogenetic tree
tree.unrooted = experiment_8@phy_tree

source(file = paste0(filepath, "/functions/pick_new_outgroup.R"))
#pick_new_outgroup result in the OTU with the longest branch 

new.outgroup = pick_new_outgroup(tree.unrooted) #Zotu897 chosen as outgroup
#Now we root the tree with this otu as root
rootedTree = ape::root(tree.unrooted, outgroup=new.outgroup, resolve.root=TRUE)
experiment_merged = phyloseq(OTU, TAX, sam_variables, rootedTree) #dataset contains 1316 taxa

```


```{r}
# makes an RDS file 
saveRDS(experiment_merged, file = paste0(filepath, "/data/zotus/2022-01-11_zOTU_raw_dataset.rds") ) 
```
